### PR TITLE
Add SST integration into gpgpusim

### PR DESCRIPTION
**Pull request by William-An**
# Add SST integration into gpgpusim

Allow SST (Structural Simulation Toolkit) to utilize GPGPUSim during simulation.

## Usage

To enable SST mode, make sure you have `-SST_mode 1` in the `gpgpusim.config` file. You should also use SST specific CUDA functions.

## Changes

1. Incorporate the previous [sst-gpgpusim-external](https://github.com/purdue-aalp/sst-gpgpusim-external) repo to current mainstream.
2. ~~Changes related to SST are encapsulated with `#ifdefine __SST__` directives with flag passed from makefile~~
3. SST specific cluster, mem interface, gpgpu_sim are created to handle SST requests.
4. Major changes were made to the following folders/files:
    1. `libcuda`: mostly with `cuda_runtime_api.cc` to add SST specific APIs as with SST we won't have access to `/proc/self/exe` to extract PTX file info.
    2. `src/gpgpu-sim/gpu-sim.cc`: for creating simt cluster and SST specific cycle function
    3. `src/gpgpu-sim/shader.cc`: for handling memory transactions with SST memory system instead of GPGPUSim's

### CODE REVIEW

**Comment by William-An** [comment]
The Jenkins run might fail a few times as we are still waiting for `devel` branch in `sst-elements` to be merged into default branch. Plus there is a bug in `sst-core` which prevent its build under balar configuration.

**Comment by William-An** [comment]
I guess there should also be a CI test for testing GPGPU-Sim with SST unit test?

**Comment by William-An** [comment]
We properly would want to have this merged some time so I don't have to point people to my personal repo in SST docs haha.

File: .github/workflows/sst_integration.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        run: |
+          source ./setup_environment
+          make -j4
+      - name: Prepare SST dependencies
+        run: |
+          apt install -y openmpi-bin openmpi-common libtool libtool-bin autoconf python3 python3-dev automake build-essential git
+      # Use personal repo for now
+      - name: Build SST-Core
+        run: |
+          git clone https://github.com/William-An/sst-core.git
```
Comment (JRPan) [inline .github/workflows/sst_integration.yml:46]: Still using personal repo?

File: libcuda/cuda_runtime_api.cc [hunk 2]
```diff
@@ -2754,11 +2890,48 @@ __host__ const char *CUDARTAPI cudaGetErrorString(cudaError_t error) {
...
+  return cudaSetupArgumentInternal(local_value, size, offset);
+}
+
 __host__ cudaError_t CUDARTAPI cudaSetupArgument(const void *arg, size_t size,
                                                  size_t offset) {
   return cudaSetupArgumentInternal(arg, size, offset);
 }
 
+// SST specific cuda apis
+// TODO Weili: Unify SST and GPGPUSim apis?
```
Comment (JRPan) [inline libcuda/cuda_runtime_api.cc]: remove TODO if not going to be fixed.

File: src/gpgpu-sim/gpu-sim.cc [hunk 3]
```diff
@@ -1132,6 +1164,27 @@ bool gpgpu_sim::active() {
...
+  if (m_config.gpu_max_cta_opt &&
+      (gpu_tot_issued_cta >= m_config.gpu_max_cta_opt))
+    return false;
+  if (m_config.gpu_max_completed_cta_opt &&
+      (gpu_completed_cta >= m_config.gpu_max_completed_cta_opt))
+    return false;
+  if (m_config.gpu_deadlock_detect && gpu_deadlock) return false;
+  for (unsigned i = 0; i < m_shader_config->n_simt_clusters; i++)
+    if (m_cluster[i]->get_not_completed() > 0) return true;
+  ;
```
Comment (JRPan) [inline src/gpgpu-sim/gpu-sim.cc]: what is this for?

File: src/gpgpu-sim/gpu-sim.cc [hunk 4]
```diff
@@ -2236,3 +2294,114 @@ const shader_core_config *gpgpu_sim::getShaderCoreConfig() {
...
+        "file and ensure n_simt_clusters"
+        "is the same as SST gpu cores.\n" &&
+        0);
+  } else {
+    printf("\nSST GPU core is equal the GPGPU-sim cores = %d\n", sst_numcores);
+  }
+}
+
+void sst_gpgpu_sim::SST_cycle() {
+  // int clock_mask = next_clock_domain();
```
Comment (JRPan) [inline src/gpgpu-sim/gpu-sim.cc]: I assume clock masks are not needed with SST?

File: src/gpgpu-sim/shader.cc [hunk 5]
```diff
@@ -4639,6 +4670,82 @@ void simt_core_cluster::icnt_inject_request_packet(class mem_fetch *mf) {
                 mf->size());
 }
```
Comment (JRPan) [inline src/gpgpu-sim/shader.cc:4629]: I guess you can factor out the stat++ part and reuse it with icnt_inject_request_packet

**Review by JRPan — CHANGES_REQUESTED** [review]
Cool. I have some questions, but not neccesary needs changes. Take a look

- Resolve the conflicts, I created PR several days ago to address gcc-10+ problem. You fixed that as well. \
- Would it make sense to move the changes in `shader.cc` and `shader.h` and create a file `sst-shader.cc`? Just saying. If it's too much trouble then I'm fine withcurrect state.

File: src/gpgpu-sim/gpu-sim.cc (hunk 3 above)
Comment (JRPan) [reply src/gpgpu-sim/gpu-sim.cc]: line 1183

File: .github/workflows/sst_integration.yml (hunk 1 above)
Comment (William-An) [reply .github/workflows/sst_integration.yml:46]: Yeah, SST side want us to merge the PR first before having my balar changes merged into their dev branch.

File: src/gpgpu-sim/gpu-sim.cc (hunk 3 above)
Comment (William-An) [reply src/gpgpu-sim/gpu-sim.cc]: I think this is a typo carried over from what Mengchi has for the initial gpgpu-sim SST repo. I will remove this.

**Comment by William-An** [comment]
I fixed most of the places except for moving the changes out of `shader.cc/.h`. Some of the code are closely mingled together, and there isn't a clean way to separate the changes unless we have a completely different chain from `gpu`->`core ctx`->`cluster`->`shader core`.

File: src/gpgpu-sim/shader.cc (hunk 5 above)
Comment (William-An) [reply src/gpgpu-sim/shader.cc:4629]: Factored out into a method `update_icnt_stats`.

File: src/gpgpu-sim/gpu-sim.cc (hunk 4 above)
Comment (William-An) [reply src/gpgpu-sim/gpu-sim.cc]: Yep, SST only need the gpgpu-sim core, other memory parts are handled by SST side.

File: libcuda/cuda_runtime_api.cc (hunk 2 above)
Comment (William-An) [reply libcuda/cuda_runtime_api.cc]: Removed all my TODOs

**Review by JRPan — APPROVED** [review]

**Pull request merged**